cluster: streamline version handling

tbg opened the pull request:
Force (and verify) an ordering between versions and make it more straightforward to add a new one.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18993)
<!-- Reviewable:end -->

nvanbenschoten commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/settings/cluster/cockroach_versions.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYx369nHJAg6MWw9z4:-KvYx369nHJAg6MWw9z5:b-a3srnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/cockroach_versions.go#L42)):*
> ```Go
> // describing what backwards-incompatible features were introduced.
> //
> // NB: when adding a version, don't forget to bump ServerVersion above (and perhaps
> ```

Is this stale?

---

*[pkg/settings/cluster/cockroach_versions.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYyUw5527_vup4mnDf:-KvYyUw5527_vup4mnDg:b82ij73) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/cockroach_versions.go#L49)):*
> ```Go
> // wiped. As a result, we recommend not bumping to a new minor version until the prior 1.X.0 release
> // has been performed.
> var versionsSingleton = makeKeyedVersions([]keyedVersion{
> ```

You could probably make this an array using the `[...]` syntax and avoid `makeKeyedVersions`. The order wouldn't matter in that case and you could just run a validation step over it. Your call.

---

*[pkg/settings/cluster/cockroach_versions.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYx5zFkU2VaMa2d6Fv:-KvYx5zFkU2VaMa2d6Fw:b-ifsjrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/cockroach_versions.go#L50)):*
> ```Go
> // has been performed.
> var versionsSingleton = makeKeyedVersions([]keyedVersion{
> 	// Add new versions here (step two of two).
> ```

These listings have different orders. Could be confusing.

---

*[pkg/settings/cluster/keyed_versions.go, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYxTS6MClguYWKNYaG:-KvYxTS6MClguYWKNYaH:bo7rqpc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/keyed_versions.go#L22)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/roachpb"
> 	"github.com/cockroachdb/cockroach/pkg/util/log"
> 	"github.com/kr/pretty"
> ```

nit: move up

---

*[pkg/settings/cluster/keyed_versions.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYyEBXWeCJTj7Z-fQb:-KvYyEBYlUScbJyL9YHF:b-md37tl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/keyed_versions.go#L48)):*
> ```Go
> func makeKeyedVersions(vs []keyedVersion) keyedVersions {
> 	ctx := context.Background()
> 	// Reverse the slice.
> ```

Maybe just avoid this and make new versions go last. Your call.

---

*[pkg/settings/cluster/settings.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYz39hijPbxV5sDSIt:-KvYz39hijPbxV5sDSIu:b-dfe8p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/settings.go#L161)):*
> ```Go
> // IsActiveVersion returns true if the features of the supplied version are active at the running
> // version.
> func (ecv *ExposedClusterVersion) IsActiveVersion(v roachpb.Version) bool {
> ```

Do we even need this anymore?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993)*
<!-- Sent from Reviewable.io -->

tbg committed:
cluster: streamline version handling

Force (and verify) an ordering between versions and make it more straightforward to add a new one.

tschottdorf commented:
Review status: 3 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/settings/cluster/cockroach_versions.go, line 42 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYx369nHJAg6MWw9z4:-KvbmZdSKSj2-kOqxQdp:b-xew2je) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/cockroach_versions.go#L42)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this stale?
</blockquote></details>

Yep, updated.

---

*[pkg/settings/cluster/cockroach_versions.go, line 49 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYyUw5527_vup4mnDf:-KvblK6h0j_6lPvKBpAv:b-h96ath) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/cockroach_versions.go#L49)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You could probably make this an array using the `[...]` syntax and avoid `makeKeyedVersions`. The order wouldn't matter in that case and you could just run a validation step over it. Your call.
</blockquote></details>

Done (can't quite use `[...]` but close enough).

---

*[pkg/settings/cluster/cockroach_versions.go, line 50 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYx5zFkU2VaMa2d6Fv:-Kvbe2Lmwqh88RLn3icW:b-8d6zar) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/cockroach_versions.go#L50)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

These listings have different orders. Could be confusing.
</blockquote></details>

Yeah, I was doing this so that adding a version would be more localized, but it's probably not worth it. Changed.

---

*[pkg/settings/cluster/keyed_versions.go, line 22 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYxTS6MClguYWKNYaG:-KvbeJekU3hRoCGgChs-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/keyed_versions.go#L22)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: move up
</blockquote></details>

Done.

---

*[pkg/settings/cluster/keyed_versions.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYyEBXWeCJTj7Z-fQb:-KvblGwVB9wkWqaDF7ke:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/keyed_versions.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Maybe just avoid this and make new versions go last. Your call.
</blockquote></details>

Done.

---

*[pkg/settings/cluster/settings.go, line 161 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-KvYz39hijPbxV5sDSIt:-KvblkN8HY9SKKOb3c2A:bd6cegr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ee0bcdb5ea45e0cf5965d99717332dbbc9b6196/pkg/settings/cluster/settings.go#L161)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we even need this anymore?
</blockquote></details>

I removed this one, but it remains on `ClusterVersion` because we sometimes test versions that don't exist.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18993#-:-Kvc-TNWGzItJWm7a-MO:bnfp4nl)*
<!-- Sent from Reviewable.io -->
